File: tensorflow_lite_support/c/test/task/vision/image_segmenter_test.cc
/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow_lite_support/c/task/vision/image_segmenter.h"

#include <string.h>

#include "tensorflow/lite/core/shims/cc/shims_test_util.h"
#include "tensorflow_lite_support/c/common.h"
#include "tensorflow_lite_support/c/task/processor/segmentation_result.h"
#include "tensorflow_lite_support/c/task/vision/core/frame_buffer.h"
#include "tensorflow_lite_support/cc/port/gmock.h"
#include "tensorflow_lite_support/cc/port/gtest.h"
#include "tensorflow_lite_support/cc/port/status_matchers.h"
#include "tensorflow_lite_support/cc/test/test_utils.h"
#include "tensorflow_lite_support/examples/task/vision/desktop/utils/image_utils.h"

#define GTEST_COUT std::cerr << "[          ] [ INFO ]"


namespace tflite {
namespace task {
namespace vision {
namespace {

using ::testing::HasSubstr;
using ::tflite::support::StatusOr;
using ::tflite::task::JoinPath;

constexpr char kTestDataDirectory[] =
    "/tensorflow_lite_support/cc/test/testdata/task/"
    "vision/";
// Quantized model.
constexpr char kDeepLabV3[] = "deeplabv3.tflite";

StatusOr<ImageData> LoadImage(const char* image_name) {
  return DecodeImageFromFile(JoinPath("./" /*test src dir*/,
                                      kTestDataDirectory, image_name));
}

class ImageSegmenterFromOptionsTest : public tflite_shims::testing::Test {};

// TEST_F(ImageClassifierFromOptionsTest, FailsWithNullOptionsAndError) {
//   TfLiteSupportError* error = nullptr;
//   TfLiteImageClassifier* image_classifier =
//       TfLiteImageClassifierFromOptions(nullptr, &error);

//   EXPECT_EQ(image_classifier, nullptr);
//   if (image_classifier) TfLiteImageClassifierDelete(image_classifier);

//   ASSERT_NE(error, nullptr);
//   EXPECT_EQ(error->code, kInvalidArgumentError);
//   EXPECT_NE(error->message, nullptr);
//   EXPECT_THAT(error->message, HasSubstr("Expected non null options"));

//   TfLiteSupportErrorDelete(error);
// }

// TEST_F(ImageClassifierFromOptionsTest, FailsWithMissingModelPath) {
//   TfLiteImageClassifierOptions options = TfLiteImageClassifierOptionsCreate();
//   TfLiteImageClassifier* image_classifier =
//       TfLiteImageClassifierFromOptions(&options, nullptr);
//   EXPECT_EQ(image_classifier, nullptr);
//   if (image_classifier) TfLiteImageClassifierDelete(image_classifier);
// }

// TEST_F(ImageClassifierFromOptionsTest, FailsWithMissingModelPathAndError) {
//   TfLiteImageClassifierOptions options = TfLiteImageClassifierOptionsCreate();

//   TfLiteSupportError* error = nullptr;
//   TfLiteImageClassifier* image_classifier =
//       TfLiteImageClassifierFromOptions(&options, &error);

//   EXPECT_EQ(image_classifier, nullptr);
//   if (image_classifier) TfLiteImageClassifierDelete(image_classifier);

//   ASSERT_NE(error, nullptr);
//   EXPECT_EQ(error->code, kInvalidArgumentError);
//   EXPECT_NE(error->message, nullptr);
//   EXPECT_THAT(error->message, HasSubstr("`base_options.model_file`"));

//   TfLiteSupportErrorDelete(error);
// }

TEST_F(ImageSegmenterFromOptionsTest, SucceedsWithModelPath) {
    GTEST_COUT << "In teest" << std::endl;
  std::string model_path =
      JoinPath("./" /*test src dir*/, kTestDataDirectory,
               kDeepLabV3);
  TfLiteImageSegmenterOptions options = TfLiteImageSegmenterOptionsCreate();
  GTEST_COUT << "Created Opt" << std::endl;

  options.base_options.model_file.file_path = model_path.data();
  GTEST_COUT << "ModelPath" << std::endl;
  TfLiteImageSegmenter* image_segmenter =
      TfLiteImageSegmenterFromOptions(&options, nullptr);
  
  GTEST_COUT << "Seggmeenteer Create" << std::endl;
  EXPECT_NE(image_segmenter, nullptr);
  TfLiteImageSegmenterDelete(image_segmenter);
}

TEST_F(ImageSegmenterFromOptionsTest, SucceedsWithNumberOfThreadsAndError) {
  std::string model_path =
      JoinPath("./" /*test src dir*/, kTestDataDirectory,
               kDeepLabV3);
  TfLiteImageSegmenterOptions options = TfLiteImageSegmenterOptionsCreate();
  options.base_options.model_file.file_path = model_path.data();
  options.base_options.compute_settings.cpu_settings.num_threads = 3;

  TfLiteSupportError* error = nullptr;
  TfLiteImageSegmenter* image_segmenter =
      TfLiteImageSegmenterFromOptions(&options, &error);

  EXPECT_NE(image_segmenter, nullptr);
  EXPECT_EQ(error, nullptr);

  if (image_segmenter) TfLiteImageSegmenterDelete(image_segmenter);
  if (error) TfLiteSupportErrorDelete(error);
}

// TEST(ImageClassifierNullClassifierClassifyTest,
//      FailsWithNullImageClassifierAndError) {
//   SUPPORT_ASSERT_OK_AND_ASSIGN(ImageData image_data, LoadImage("burger-224.png"));

//   TfLiteSupportError* error = nullptr;
//   TfLiteClassificationResult* classification_result =
//       TfLiteImageClassifierClassify(nullptr, nullptr, &error);

//   ImageDataFree(&image_data);

//   EXPECT_EQ(classification_result, nullptr);
//   if (classification_result)
//     TfLiteClassificationResultDelete(classification_result);

//   ASSERT_NE(error, nullptr);
//   EXPECT_EQ(error->code, kInvalidArgumentError);
//   EXPECT_NE(error->message, nullptr);
//   EXPECT_THAT(error->message, HasSubstr("Expected non null image classifier"));

//   TfLiteSupportErrorDelete(error);
// }

class ImageSegmenterSegmentTest : public tflite_shims::testing::Test {
 protected:
  void SetUp() override {
    std::string model_path =
        JoinPath("./" /*test src dir*/, kTestDataDirectory,
                 kDeepLabV3);

    TfLiteImageSegmenterOptions options = TfLiteImageSegmenterOptionsCreate();
    options.base_options.model_file.file_path = model_path.data();
    image_classifier = TfLiteImageSegmenterFromOptions(&options, nullptr);
    ASSERT_NE(image_classifier, nullptr);
  }

  void TearDown() override { TfLiteImageSegmenterDelete(image_classifier); }
  TfLiteImageSegmenter* image_classifier;
};

TEST_F(ImageSegmenterSegmentTest, SucceedsWithImageData) {
  SUPPORT_ASSERT_OK_AND_ASSIGN(ImageData image_data, LoadImage("burger_crop.jpg"));

  TfLiteFrameBuffer frame_buffer = {
      .format = kRGB,
      .orientation = kTopLeft,
      .dimension = {.width = image_data.width, .height = image_data.height},
      .buffer = image_data.pixel_data};

  TfLiteSegmentationResult* segmentation_result = TfLiteImageSegmenterSegment(image_classifier, &frame_buffer, nullptr);

  ImageDataFree(&image_data);
  GTEST_COUT << "Done" << std::endl;

  // ASSERT_NE(classification_result, nullptr);
  // EXPECT_GE(classification_result->size, 1);
  // EXPECT_NE(classification_result->classifications, nullptr);
  // EXPECT_GE(classification_result->classifications->size, 1);
  // EXPECT_NE(classification_result->classifications->categories, nullptr);
  // EXPECT_EQ(strcmp(classification_result->classifications->categories[0].label,
  //                  "cheeseburger"),
  //           0);
  // EXPECT_GE(classification_result->classifications->categories[0].score, 0.90);

  // TfLiteClassificationResultDelete(classification_result);
}

// TEST_F(ImageClassifierClassifyTest, FailsWithNullFrameBufferAndError) {
//   SUPPORT_ASSERT_OK_AND_ASSIGN(ImageData image_data, LoadImage("burger-224.png"));

//   TfLiteSupportError* error = nullptr;
//   TfLiteClassificationResult* classification_result =
//       TfLiteImageClassifierClassify(image_classifier, nullptr, &error);

//   ImageDataFree(&image_data);

//   EXPECT_EQ(classification_result, nullptr);
//   if (classification_result)
//     TfLiteClassificationResultDelete(classification_result);

//   ASSERT_NE(error, nullptr);
//   EXPECT_EQ(error->code, kInvalidArgumentError);
//   EXPECT_NE(error->message, nullptr);
//   EXPECT_THAT(error->message, HasSubstr("Expected non null frame buffer"));

//   TfLiteSupportErrorDelete(error);
// }

// TEST_F(ImageClassifierClassifyTest, FailsWithNullImageDataAndError) {
//   SUPPORT_ASSERT_OK_AND_ASSIGN(ImageData image_data, LoadImage("burger-224.png"));

//   TfLiteFrameBuffer frame_buffer = {.format = kRGB, .orientation = kTopLeft};

//   TfLiteSupportError* error = nullptr;
//   TfLiteClassificationResult* classification_result =
//       TfLiteImageClassifierClassify(image_classifier, &frame_buffer, &error);

//   ImageDataFree(&image_data);

//   EXPECT_EQ(classification_result, nullptr);
//   if (classification_result)
//     TfLiteClassificationResultDelete(classification_result);

//   ASSERT_NE(error, nullptr);
//   EXPECT_EQ(error->code, kInvalidArgumentError);
//   EXPECT_NE(error->message, nullptr);
//   EXPECT_THAT(error->message, HasSubstr("Invalid stride information"));

//   TfLiteSupportErrorDelete(error);
// }

// TEST_F(ImageClassifierClassifyTest, SucceedsWithRoiWithinImageBounds) {
//   SUPPORT_ASSERT_OK_AND_ASSIGN(ImageData image_data, LoadImage("burger-224.png"));

//   TfLiteFrameBuffer frame_buffer = {
//       .format = kRGB,
//       .orientation = kTopLeft,
//       .dimension = {.width = image_data.width, .height = image_data.height},
//       .buffer = image_data.pixel_data};

//   TfLiteBoundingBox bounding_box = {
//       .origin_x = 0, .origin_y = 0, .width = 100, .height = 100};
//   TfLiteSupportError* error = nullptr;
//   TfLiteClassificationResult* classification_result =
//       TfLiteImageClassifierClassifyWithRoi(image_classifier, &frame_buffer,
//                                            &bounding_box, &error);

//   ImageDataFree(&image_data);

//   ASSERT_NE(classification_result, nullptr);
//   EXPECT_GE(classification_result->size, 1);
//   EXPECT_NE(classification_result->classifications, nullptr);
//   EXPECT_GE(classification_result->classifications->size, 1);
//   EXPECT_NE(classification_result->classifications->categories, nullptr);
//   EXPECT_EQ(strcmp(classification_result->classifications->categories[0].label,
//                    "bagel"),
//             0);
//   EXPECT_GE(classification_result->classifications->categories[0].score, 0.30);

//   TfLiteClassificationResultDelete(classification_result);
// }

// TEST_F(ImageClassifierClassifyTest, FailsWithRoiOutsideImageBoundsAndError) {
//   SUPPORT_ASSERT_OK_AND_ASSIGN(ImageData image_data, LoadImage("burger-224.png"));

//   TfLiteFrameBuffer frame_buffer = {
//       .format = kRGB,
//       .orientation = kTopLeft,
//       .dimension = {.width = image_data.width, .height = image_data.height},
//       .buffer = image_data.pixel_data};

//   TfLiteBoundingBox bounding_box = {
//       .origin_x = 0, .origin_y = 0, .width = 250, .height = 250};
//   TfLiteSupportError* error = nullptr;
//   TfLiteClassificationResult* classification_result =
//       TfLiteImageClassifierClassifyWithRoi(image_classifier, &frame_buffer,
//                                            &bounding_box, &error);

//   ImageDataFree(&image_data);

//   EXPECT_EQ(classification_result, nullptr);
//   if (classification_result)
//     TfLiteClassificationResultDelete(classification_result);

//   ASSERT_NE(error, nullptr);
//   EXPECT_EQ(error->code, kInvalidArgumentError);
//   EXPECT_NE(error->message, nullptr);
//   EXPECT_THAT(error->message, HasSubstr("Invalid crop coordinates"));

//   TfLiteSupportErrorDelete(error);
// }

// TEST(ImageClassifierWithUserDefinedOptionsClassifyTest,
//      SucceedsWithClassNameDenyList) {
//   char* denylisted_label_name = (char*)"cheeseburger";
//   std::string model_path =
//       JoinPath("./" /*test src dir*/, kTestDataDirectory,
//                kMobileNetQuantizedWithMetadata);

//   TfLiteImageClassifierOptions options = TfLiteImageClassifierOptionsCreate();
//   options.base_options.model_file.file_path = model_path.data();

//   char* label_denylist[12] = {denylisted_label_name};
//   options.classification_options.label_denylist.list = label_denylist;
//   options.classification_options.label_denylist.length = 1;

//   TfLiteImageClassifier* image_classifier =
//       TfLiteImageClassifierFromOptions(&options, nullptr);
//   ASSERT_NE(image_classifier, nullptr);

//   SUPPORT_ASSERT_OK_AND_ASSIGN(ImageData image_data, LoadImage("burger-224.png"));

//   TfLiteFrameBuffer frame_buffer = {
//       .format = kRGB,
//       .orientation = kTopLeft,
//       .dimension = {.width = image_data.width, .height = image_data.height},
//       .buffer = image_data.pixel_data};

//   TfLiteClassificationResult* classification_result =
//       TfLiteImageClassifierClassify(image_classifier, &frame_buffer, nullptr);

//   ImageDataFree(&image_data);
//   if (image_classifier) TfLiteImageClassifierDelete(image_classifier);

//   ASSERT_NE(classification_result, nullptr);
//   EXPECT_GE(classification_result->size, 1);
//   EXPECT_NE(classification_result->classifications, nullptr);
//   EXPECT_GE(classification_result->classifications->size, 1);
//   EXPECT_NE(classification_result->classifications->categories, nullptr);
//   EXPECT_NE(strcmp(classification_result->classifications->categories[0].label,
//                    denylisted_label_name),
//             0);

//   TfLiteClassificationResultDelete(classification_result);
// }

// TEST(ImageClassifierWithUserDefinedOptionsClassifyTest,
//      SucceedsWithClassNameAllowList) {
//   char* allowlisted_label_name = (char*)"cheeseburger";
//   std::string model_path =
//       JoinPath("./" /*test src dir*/, kTestDataDirectory,
//                kMobileNetQuantizedWithMetadata)
//           .data();

//   TfLiteImageClassifierOptions options = TfLiteImageClassifierOptionsCreate();
//   options.base_options.model_file.file_path = model_path.data();

//   char* label_allowlist[12] = {allowlisted_label_name};
//   options.classification_options.label_allowlist.list = label_allowlist;
//   options.classification_options.label_allowlist.length = 1;

//   TfLiteImageClassifier* image_classifier =
//       TfLiteImageClassifierFromOptions(&options, nullptr);
//   ASSERT_NE(image_classifier, nullptr);

//   SUPPORT_ASSERT_OK_AND_ASSIGN(ImageData image_data, LoadImage("burger-224.png"));

//   TfLiteFrameBuffer frame_buffer = {
//       .format = kRGB,
//       .orientation = kTopLeft,
//       .dimension = {.width = image_data.width, .height = image_data.height},
//       .buffer = image_data.pixel_data};

//   TfLiteClassificationResult* classification_result =
//       TfLiteImageClassifierClassify(image_classifier, &frame_buffer, nullptr);

//   ImageDataFree(&image_data);
//   if (image_classifier) TfLiteImageClassifierDelete(image_classifier);

//   ASSERT_NE(classification_result, nullptr);
//   EXPECT_GE(classification_result->size, 1);
//   EXPECT_NE(classification_result->classifications, nullptr);
//   EXPECT_GE(classification_result->classifications->size, 1);
//   EXPECT_NE(classification_result->classifications->categories, nullptr);
//   EXPECT_EQ(strcmp(classification_result->classifications->categories[0].label,
//                    allowlisted_label_name),
//             0);

//   TfLiteClassificationResultDelete(classification_result);
// }

}  // namespace
}  // namespace vision
}  // namespace task
}  // namespace tflite

File: tensorflow_lite_support/python/test/task/vision/image_segmenter_test.py
# Copyright 2022 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Tests for image_segmenter."""

import enum

from absl.testing import parameterized
import numpy as np
import tensorflow as tf

from tensorflow_lite_support.python.task.core.proto import base_options_pb2
from tensorflow_lite_support.python.task.processor.proto import segmentation_options_pb2
from tensorflow_lite_support.python.task.vision import image_segmenter
from tensorflow_lite_support.python.task.vision.core import tensor_image
from tensorflow_lite_support.python.test import test_util

_BaseOptions = base_options_pb2.BaseOptions
_OutputType = segmentation_options_pb2.SegmentationOptions.OutputType
_ImageSegmenter = image_segmenter.ImageSegmenter
_ImageSegmenterOptions = image_segmenter.ImageSegmenterOptions

_MODEL_FILE = 'deeplabv3.tflite'
_IMAGE_FILE = 'segmentation_input_rotation0.jpg'
_SEGMENTATION_FILE = 'segmentation_golden_rotation0.png'
_EXPECTED_COLORED_LABELS = [{
    'r': 0,
    'g': 0,
    'b': 0,
    'class_name': 'background'
}, {
    'r': 128,
    'g': 0,
    'b': 0,
    'class_name': 'aeroplane'
}, {
    'r': 0,
    'g': 128,
    'b': 0,
    'class_name': 'bicycle'
}, {
    'r': 128,
    'g': 128,
    'b': 0,
    'class_name': 'bird'
}, {
    'r': 0,
    'g': 0,
    'b': 128,
    'class_name': 'boat'
}, {
    'r': 128,
    'g': 0,
    'b': 128,
    'class_name': 'bottle'
}, {
    'r': 0,
    'g': 128,
    'b': 128,
    'class_name': 'bus'
}, {
    'r': 128,
    'g': 128,
    'b': 128,
    'class_name': 'car'
}, {
    'r': 64,
    'g': 0,
    'b': 0,
    'class_name': 'cat'
}, {
    'r': 192,
    'g': 0,
    'b': 0,
    'class_name': 'chair'
}, {
    'r': 64,
    'g': 128,
    'b': 0,
    'class_name': 'cow'
}, {
    'r': 192,
    'g': 128,
    'b': 0,
    'class_name': 'dining table'
}, {
    'r': 64,
    'g': 0,
    'b': 128,
    'class_name': 'dog'
}, {
    'r': 192,
    'g': 0,
    'b': 128,
    'class_name': 'horse'
}, {
    'r': 64,
    'g': 128,
    'b': 128,
    'class_name': 'motorbike'
}, {
    'r': 192,
    'g': 128,
    'b': 128,
    'class_name': 'person'
}, {
    'r': 0,
    'g': 64,
    'b': 0,
    'class_name': 'potted plant'
}, {
    'r': 128,
    'g': 64,
    'b': 0,
    'class_name': 'sheep'
}, {
    'r': 0,
    'g': 192,
    'b': 0,
    'class_name': 'sofa'
}, {
    'r': 128,
    'g': 192,
    'b': 0,
    'class_name': 'train'
}, {
    'r': 0,
    'g': 64,
    'b': 128,
    'class_name': 'tv'
}]
_MASK_MAGNIFICATION_FACTOR = 10
_MATCH_PIXELS_THRESHOLD = 0.01


def _create_segmenter_from_options(base_options, **segmentation_options):
  segmentation_options = segmentation_options_pb2.SegmentationOptions(
      **segmentation_options)
  options = _ImageSegmenterOptions(
      base_options=base_options, segmentation_options=segmentation_options)
  segmenter = _ImageSegmenter.create_from_options(options)
  return segmenter


class ModelFileType(enum.Enum):
  FILE_CONTENT = 1
  FILE_NAME = 2


class ImageSegmenterTest(parameterized.TestCase, tf.test.TestCase):

  def setUp(self):
    super().setUp()
    self.test_image_path = test_util.get_test_data_path(_IMAGE_FILE)
    self.test_seg_path = test_util.get_test_data_path(_SEGMENTATION_FILE)
    self.model_path = test_util.get_test_data_path(_MODEL_FILE)

  def test_create_from_file_succeeds_with_valid_model_path(self):
    # Creates with default option and valid model file successfully.
    segmenter = _ImageSegmenter.create_from_file(self.model_path)
    self.assertIsInstance(segmenter, _ImageSegmenter)

  def test_create_from_options_succeeds_with_valid_model_path(self):
    # Creates with options containing model file successfully.
    base_options = _BaseOptions(file_name=self.model_path)
    options = _ImageSegmenterOptions(base_options=base_options)
    segmenter = _ImageSegmenter.create_from_options(options)
    self.assertIsInstance(segmenter, _ImageSegmenter)

  def test_create_from_options_fails_with_invalid_model_path(self):
    # Invalid empty model path.
    with self.assertRaisesRegex(
        ValueError,
        r"ExternalFile must specify at least one of 'file_content', "
        r"'file_name' or 'file_descriptor_meta'."):
      base_options = _BaseOptions(file_name='')
      options = _ImageSegmenterOptions(base_options=base_options)
      _ImageSegmenter.create_from_options(options)

  def test_create_from_options_succeeds_with_valid_model_content(self):
    # Creates with options containing model content successfully.
    with open(self.model_path, 'rb') as f:
      base_options = _BaseOptions(file_content=f.read())
      options = _ImageSegmenterOptions(base_options=base_options)
      segmenter = _ImageSegmenter.create_from_options(options)
      self.assertIsInstance(segmenter, _ImageSegmenter)

  @parameterized.parameters(
      (ModelFileType.FILE_NAME, _EXPECTED_COLORED_LABELS),
      (ModelFileType.FILE_CONTENT, _EXPECTED_COLORED_LABELS))
  def test_segment_model(self, model_file_type, expected_colored_labels):
    # Creates segmenter.
    if model_file_type is ModelFileType.FILE_NAME:
      base_options = _BaseOptions(file_name=self.model_path)
    elif model_file_type is ModelFileType.FILE_CONTENT:
      with open(self.model_path, 'rb') as f:
        model_content = f.read()
      base_options = _BaseOptions(file_content=model_content)
    else:
      # Should never happen
      raise ValueError('model_file_type is invalid.')

    segmenter = _create_segmenter_from_options(base_options)

    # Loads image.
    image = tensor_image.TensorImage.create_from_file(self.test_image_path)

    # Performs image segmentation on the input.
    segmentation = segmenter.segment(image).segmentation[0]
    colored_labels = segmentation.colored_labels

    # Check if the sizes of the result and expected colored labels are the same.
    self.assertEqual(
        len(colored_labels), len(expected_colored_labels),
        'Number of colored labels do not match.')

    # Comparing results.
    for index in range(len(expected_colored_labels)):
      for key in expected_colored_labels[index].keys():
        self.assertEqual(
            getattr(colored_labels[index], key),
            expected_colored_labels[index][key])

  def test_segmentation_category_mask(self):
    """Check if category mask matches with ground truth."""
    # Creates segmenter.
    base_options = _BaseOptions(file_name=self.model_path)
    segmenter = _create_segmenter_from_options(
        base_options, output_type=_OutputType.CATEGORY_MASK)

    # Loads image.
    image = tensor_image.TensorImage.create_from_file(self.test_image_path)

    # Performs image segmentation on the input.
    segmentation = segmenter.segment(image).segmentation[0]
    result_pixels = np.array(bytearray(segmentation.category_mask))

    # Loads ground truth segmentation file.
    gt_segmentation = tensor_image.TensorImage.create_from_file(
        self.test_seg_path)
    gt_segmentation_array = gt_segmentation.buffer
    gt_segmentation_shape = gt_segmentation_array.shape
    num_pixels = gt_segmentation_shape[0] * gt_segmentation_shape[1]
    ground_truth_pixels = gt_segmentation_array.flatten()

    self.assertEqual(
        len(result_pixels), len(ground_truth_pixels),
        'Segmentation mask size does not match the ground truth mask size.')

    inconsistent_pixels = 0

    for index in range(num_pixels):
      inconsistent_pixels += (
          result_pixels[index] * _MASK_MAGNIFICATION_FACTOR !=
          ground_truth_pixels[index])

    self.assertLessEqual(
        inconsistent_pixels / num_pixels, _MATCH_PIXELS_THRESHOLD,
        f'Number of pixels in the candidate mask differing from that of the '
        f'ground truth mask exceeds {_MATCH_PIXELS_THRESHOLD}.')

  def test_segmentation_confidence_mask_matches_category_mask(self):
    """Check if the confidence mask matches with the category mask."""
    # Create BaseOptions from model file.
    base_options = _BaseOptions(file_name=self.model_path)

    # Loads image.
    image = tensor_image.TensorImage.create_from_file(self.test_image_path)

    # Run segmentation on the model in CATEGORY_MASK mode.
    segmenter = _create_segmenter_from_options(
        base_options, output_type=_OutputType.CATEGORY_MASK)

    # Performs image segmentation on the input and gets the category mask.
    segmentation = segmenter.segment(image).segmentation[0]
    category_mask = np.array(bytearray(segmentation.category_mask))

    # Run segmentation on the model in CONFIDENCE_MASK mode.
    segmenter = _create_segmenter_from_options(
        base_options, output_type=_OutputType.CONFIDENCE_MASK)

    # Performs image segmentation on the input again.
    segmentation = segmenter.segment(image).segmentation[0]

    # Gets the list of confidence masks and colored_labels.
    confidence_masks = segmentation.confidence_masks.confidence_mask
    colored_labels = segmentation.colored_labels

    # Check if confidence mask shape is correct.
    self.assertEqual(
        len(confidence_masks), len(colored_labels),
        'Number of confidence masks must match with number of categories.')

    # Gather the confidence masks in a single array `confidence_mask_array`.
    confidence_mask_array = np.array([
        confidence_masks[index].value for index in range(len(confidence_masks))
    ])

    # Compute the category mask from the created confidence mask.
    calculated_category_mask = np.argmax(confidence_mask_array, axis=0)
    self.assertListEqual(
        calculated_category_mask.tolist(), category_mask.tolist(),
        'Confidence mask does not match with the category mask.')


if __name__ == '__main__':
  tf.test.main()
